CONTRIBUTORS.md

```diff
@@ -154,6 +154,93 @@ All contributions should use the same whitespacing as the rest of the project.
 Moreover, Pull requests where a large number of changes only deal with whitespace will be
 rejected.
 
+# Comments
+Comments are very important. Not only are they are an aid to the reader, but in many
+cases the comments specify the contract against which the code is evaluated.
+Without a clear contract for a function, for example, it is *impossible* to determine
+whether the function is implemented correctly.
+
```

Comment:
```suggestion
Here are high level rules for creating comments:
- Use markdown to format comments, and use the triple backlash "///"
- Start with a high-level description of the function, adding more sentences with details if necessary.
- For some functions, you might want to add a markdown section for panics, errors, none results, or for examples.
- Rust has explicit typing, so it most cases explicit documentation for input and outputs of functions is not necessary. However, if there is any ambiguity, document it in the function comment. 
```

Reply:
I agree with adding @pavitthrap's suggestion, but should it read `Here are high level rules for creating function comments` (because it doesn't seem to apply to things like inline comments, etc.)?

Reply:
While what you have here looks good, I want to also make sure that our contributor guide doesn't become a form of bureaucratic gatekeeping that generates unnecessary work for contributors and results in low-entropy documentation for maintainers, all just to fulfill a PR checkbox.  While I like what you have here, I think there should be a bullet to the effect of "use your best judgement to document that which is not obvious."  It's up to the PR reviewers to ask for more comments if they don't understand how the code works, or if they feel that a 3rd party reviewer wouldn't readily understand.  But simple functions whose name describes what it does (e.g. getters and setters) I don't think need comment blocks.

Reply:
> The goal should be to not put nothing,

I guess this is a difference between our personal philosophies, but I think that this should be *exactly* the goal.  The goal for *code* should be to write it so clearly that it's self-documenting -- the code itself should tell you what's happening.  This is obviously not realizable in practice, but it's an ideal I think we should hold.

The purpose of documentation, then, would be to explain that which cannot be immediately deduced from the code itself.  Namely, the code describes *what* happens, and the documentation describes what *should* happen.  Documentation, then, includes higher-level information like where a particular method or struct fits into the overall system architecture, and what its intended use-cases are and are not.  This would also include module-level documentation, as well as (in our case) documentation about what is and is not consensus-critical code.

You can see this in my code, for example.  I tend to write more free-form documentation blocks that try to describe the code's intended purpose and behavior in the overall system, and I generally eschew things like item-by-item documentation for method arguments and struct members (seeking instead to give them names and types that describe their purpose).  I also expect the reader to have some familiarity with the SIPs.  But in all cases, I absolutely expect the reader to read and comprehend the code in addition to reading the documentation.  If the code is so non-obvious that it needs lots of documentation to explain, then it's a bug in the code -- the code should be made more comprehensible.

Reply:
> I guess this is a difference between our personal philosophies, but I think that this should be exactly the goal. The goal for code should be to write it so clearly that it's self-documenting -- the code itself should tell you what's happening. This is obviously not realizable in practice, but it's an ideal I think we should hold.

The problem with this approach is two-fold:

1. There's a lot of room for difference in what people think is clear or unclear. For the example of function arguments, there are dozens, if not more, examples in the codebase of function arguments that are named such that they describe sufficiently what they are (e.g., `clarity_tx`), but not really at all indicating what their relationship to the function's calling contract is. Writing rustdoc comments for function arguments is not terribly difficult, and something like "`clarity_tx` should be a transaction opened to chain tip" helps a lot to indicate how a given function's lookup is affected by the open block, etc.
2. @gregorycoppola's points about helping to review should not be understated -- code _is only_ capable of describing what it does. Not what it should do, or why it does. You need look no further than our own PRs to see that a lot of explanatory text occurs _beyond_ the code. Having that content all captured in comments enables future readers to see that explanation in the code, rather than it being lost to time in a merged PR. For a current reviewer, they can see a described calling contract or description of a parameter, struct, global, etc., and know to look for a correct implementation of that description _and_ sufficient testing.

I agree with you, @jcnelson, that the described guidelines here may encourage some needless verbosity in comments, but the burden of writing comments is not significant, and given that a major bottleneck in development currently is PR review time, I think the project would benefit from experimenting with forcing-functions that potentially trade-off some increased development burden for improvements to the review process.

Reply:
Here's a pithy description of the problem I'm trying to avoid.

What I want to see in PR reviews:

* Hi $SUBMITTER, I don't quite understand what's going on here. Can you add a comment here to describe why this code does what it does?

What I do *not* want to see in PR reviews:

* Hi $SUBMITTER, your comment here exceeds 80 columns. Please fix.
* Hi $SUBMITTER, you forgot to add a 10-line comment block for this 1-line self-explanatory getter
* Hi $SUBMITTER, you didn't add a comment for this struct field member even though its usage is obvious

The above sorts of nitpicking not only wastes both the reviewer's and submitter's time, but also fail to actually solve the problem that documentation solves: provide clarity to the reader as to what the code is supposed to be doing.

> given that a major bottleneck in development currently is PR review time

I agree.  In the spirit of addressing this, I'm trying to point out that opening the domain of discourse for code review to potentially hundreds of documentation nitpicking comments like the above will make this worse.  Can I get an agreement here that this will *not* happen?

Reply:
Like, I don't want this style guide to require us to spend non-trivial amounts of cognitive overhead policing other peoples' comments, especially if the requested changes don't actually add any value.

Reply:
I think the notion of "nitpicking" boils down to a situation where two people differ on what they think is important in code style.

This what an explicitly agreed on style guide is meant to address. 

If something is considered important enough to be in the style guide, then insisting on it isn't nitpicking. And if something isn't covered by the style guide, then insisting on it *is* nitpicking.

Individuals might disagree with individual points in the style guide, just like they could disagree on any product decision, but at least everyone agrees on what to ask for and what to include in PR's.

Reply:
I guess some of the material I have been putting in comments is redundant, e.g.:

```
impl<T: MarfTrieId> MARF<T> {
    /// Reopen this MARF with readonly storage.
    ///
    /// Returns Err if:
    ///   1) This class is already in the process of writing.
    ///   2) A new underlying SQLite database connection cannot be established.
    pub fn reopen_readonly(&self) -> Result<MARF<T>, Error> {
```

You could make a rule like `don't restate the sentence in the function name; assume the user has read it, and interpreted it as English`, in which case the above could be reduced to:

```
impl<T: MarfTrieId> MARF<T> {
    /// Returns Err if:
    ///   1) This class is already in the process of writing.
    ///   2) A new underlying SQLite database connection cannot be established.
    pub fn reopen_readonly(&self) -> Result<MARF<T>, Error> {
```

Reply:
I'm less concerned about the case where an individual contributor chooses to go above and beyond the documentation requirements.  This example here is fine by me -- if I were a reviewer, I would be happy with either docstring, since in both cases I come away from reading the docstring and function signature understanding the code's contract.

What I'm more concerned about is whether or not the contributor guide will require us to spend more time worrying about aspects of documentation that *aren't adding value*.  For example, if the contributor guide mandated that each function have a sentence describing what the function does (even if it's obvious in the name), then (a) the contributor would be required to add unnecessary documentation and (b) the reviewers would be obliged to make sure the contributor does this for each function they add, even if it's not necessary.

I do think that reviewers should ask for more documentation if they believe it's warranted, and I agree with @kantai that the burden of adding more documentation isn't significant on the contributor.  However, I don't think that reviewers should have to spend an inordinate amount of time verifying that the documentation is consistent with a contributor guide.  **As long as the reviewer can read the docstring and function or struct signature and correctly comprehend the code's contract, then the documentation as given should be considered sufficient.**

Reply:
Clarifying from offline sync:

* Reviewers should ask for comments and documentation where appropriate.  Such asks are PR-blocking.
* Reviewers are not required to rigorously check that a comment block adheres to a particular formatting standard, although self-imposed adherence to rustdoc is encouraged.

Reply:
What about the *content* of comments though? 

Should we say that the **contracts** of functions should be included by default? We can make a rule like "don't restate redundant information from the declaration in the contract", to address the redundancy concern.

But, to me, the reviewing process would involve such steps as:
* compare a function's implementation to its contract
* compare a function's tests to its contract
* compare the contracts of smaller functions, to see how they build into larger functions

I guess I'm curious what your reviewing algorithm is if it doesn't involve the function contracts? Or, maybe you have such a model of the overall codebase and the work being done that you implicitly know what the contract of each method is even if its not documented? @jcnelson 

I mean, the concept of a function having a "bug" sort of implies to me that the contract exists somewhere (either in someone's mind, or PR description, if not the comments). If the definition itself were the contract, there would be no such thing as a bug, but we do find bugs. (Is this getting too philosophical?)

So, if functions do have contracts that exist independently of the implementations, I think it helps the reviewers and the author verify accuracy if we state the contracts explicitly in the comments. Cf., @kantai reviewing my comments on the ReadOnlyChecker.